### PR TITLE
EEID support for Windows

### PR DESCRIPTION
**Pull request by wintersteiger**
This adds EEID support for Windows.

There are a couple of bits that need discussion:

- [x] BCrypt doesn't give us access to the internal SHA256 hash state in a nice way; so this code contains magic offets. I'm getting in touch with the BCrypt team to see whether we can do something about that.
- [x] In the `clangw` script, I remove `-std:c11`, because MSVC claims that we would have to use `__asm` instead of `asm`. I'm not sure whether that's correct or not, but if so, we would have to change a few other macros. 
- [x] MSVC has a prototype for `memcpy` that does not contain the `restrict` keyword, so it reports a mismatch, see change in `corelibc/bits/defs.h`.

Also adds attestation plugin support to the host_verify sample.

### CODE REVIEW

**Comment by wintersteiger** [comment]
bors try

File: common/sgx/eeid.c
Comment (CodeMonkeyLeet) [inline common/sgx/eeid.c:456]: As a general design suggestion, instead of replicating `_verify_signature` 3x for different build configuration macros, consider defining a new `oe_rsa_public_key_from_modulus` method in openenclave/internal/rsa.h, and implementing that in the rsa.c for each of the supported crypto libs; verification should then have a common code path (load the key and verify signature through OE wrapper crypto functions) which reduces repetition of code and having to navigate the preprocessor options in this file. 

For reference, see the existing `oe_rsa_public_key_get_modulus/expoenent` methods, which are effectively symmetric with what you need here.

File: samples/host_verify/host.c
Comment (CodeMonkeyLeet) [inline samples/host_verify/host.c:255]: Can you also update the samples/host_verify/README.md for changes in the command line and expected usage?

@yentsanglee can you also review it from the attestation samples perspective?

File: common/sgx/eeid.c
Comment (dthaler) [reply common/sgx/eeid.c:116]: Vice versa.  
include\openenclave\host.h already has:
```
#ifndef _WIN32
#define sscanf_s sscanf   
#endif
```
enclave.h can do the same.

File: common/sgx/eeid.c
Comment (dthaler) [inline common/sgx/eeid.c:465]: https://github.com/openenclave/openenclave/blob/master/docs/Community/DevelopmentGuide.md#naming-conventions says:

> Use lower_snake_case for variable, member/field, and function names.

That means no Hungarian.

File: host/crypto/bcrypt/sha.c
Comment (dthaler) [inline host/crypto/bcrypt/sha.c:16]: Use lower_snake_case for variable, member/field, and function names.

File: host/crypto/bcrypt/sha.c
Comment (dthaler) [inline host/crypto/bcrypt/sha.c:33]: Use lower_snake_case for variable, member/field, and function names.

File: host/crypto/bcrypt/sha.c
Comment (dthaler) [inline host/crypto/bcrypt/sha.c:127]:
```suggestion
    uint32_t* hashed_count)
```
Avoid abbreviations like 'num'

File: host/crypto/bcrypt/sha.c
Comment (dthaler) [inline host/crypto/bcrypt/sha.c:158]:
```suggestion
    const uint32_t* hashed_count)
```

File: samples/host_verify/host.c
Comment (dthaler) [inline samples/host_verify/host.c:155]: Why is line 155 needed?  Shouldn't this be done automatically on first call to oe_verify_evidence()?
We should keep samples free of unnecessary code, as much as possible.

File: tools/oesign/oedump_eeid.c
Comment (wintersteiger) [reply tools/oesign/oedump_eeid.c:81]: Thanks, I didn't think there would be a solution for this problem yet!

File: scripts/clangw
Comment (wintersteiger) [reply scripts/clangw:62]: That's right, it's a combined cmake+clang problem as cmake adds `-std:c11`, see here: #3474

File: samples/host_verify/CMakeLists.txt
Comment (wintersteiger) [reply samples/host_verify/CMakeLists.txt:68]: Done. I'm not sure why I listed `oe_verify_report_internal` here; when I just tried it, it worked fine without `oehost`.

File: common/sgx/eeid.c
Comment (wintersteiger) [reply common/sgx/eeid.c:116]: Thanks, fixed!

File: common/sgx/eeid.c
Comment (wintersteiger) [reply common/sgx/eeid.c:465]: Thanks, that made everything a bit less complicated!

File: host/crypto/bcrypt/sha.c
Comment (wintersteiger) [reply host/crypto/bcrypt/sha.c]: Done. Cleaned it up and added some more comments.

File: host/crypto/bcrypt/sha.c
Comment (wintersteiger) [reply host/crypto/bcrypt/sha.c:138]: Yes, it could get nasty if the memory layout changes, but I would hope that doesn't happen frequently.

I have no idea about the certification of any of the libraries. I think there's an old version of OpenSSL that does have certification, but I don't know if that covers everything we need. Conversely, it seems to me that BCrypt doesn't actually run in "FIPS mode" by default.

File: tests/tools/oecert/enc/enc.cpp
Comment (dthaler) [inline tests/tools/oecert/enc/enc.cpp:133]: Same question about why this initialization is needed, it should be done inside oe_get_evidence() if not already done.

File: tests/tools/oecert/host/host.cpp
Comment (dthaler) [inline tests/tools/oecert/host/host.cpp:34]: Avoid abbreviations like "gen".  Yes I see it was already that was in the other two lines so this is a non-blocking suggestion to expand it throughout this file.

File: tests/tools/oecert/enc/enc.cpp
Comment (wintersteiger) [reply tests/tools/oecert/enc/enc.cpp:133]: Yes, I think you're right, so I added that to oecert so I can test it. Is the idea that all plugins will be initialized automatically or only the SGX ones (but, e.g., not the EEID plugin)?

File: tests/tools/oecert/host/host.cpp
Comment (wintersteiger) [reply tests/tools/oecert/host/host.cpp]: Sure, no problem!

File: tests/tools/oecert/enc/enc.cpp
Comment (wintersteiger) [reply tests/tools/oecert/enc/enc.cpp:134]: Turns out that's not the case; I need both `oe_verifier_initialize` and `oe_attester_initialize`, without them I get `OE_NOT_FOUND` errors. I think the design document says that the plugins should be initialized/registered automatically, but currently only `oe_verify_report` does so and `oe_verify_evidence` doesn't.

File: samples/host_verify/host.c
Comment (wintersteiger) [reply samples/host_verify/host.c:155]: (see below)

File: samples/host_verify/host.c
Comment (wintersteiger) [reply samples/host_verify/host.c:255]: Sure, README has been updated.

File: common/sgx/eeid.c
Comment (wintersteiger) [reply common/sgx/eeid.c:455]: Oh yes, you're absolutely right, thanks for the suggestion!

Comment (wintersteiger) [reply common/sgx/eeid.c]: (Done)

**Comment by wintersteiger** [comment]
bors try

**Comment by wintersteiger** [comment]
bors try

**Comment by wintersteiger** [comment]
bors retry

**Comment by wintersteiger** [comment]
bors try

**Comment by wintersteiger** [comment]
bors try

**Comment by wintersteiger** [comment]
bors try

**Comment by wintersteiger** [comment]
bors try

**Comment by wintersteiger** [comment]
bors try

File: samples/host_verify/host.c
Comment (yentsanglee) [reply samples/host_verify/host.c:274]: Endorsement_file is an optional for -r and -v:

        "Usage:\n  %s -r <report_file> [-e <endorsement_file>]\n"
                      "  %s -v <evidence_file> "[-e <endorsement_file>]\n"
                      "  %s -c <certificate_file>\n",
                      program_name,
                      program_name,
                      program_name);

File: tests/tools/oecert/enc/enc.cpp
Comment (yentsanglee) [reply tests/tools/oecert/enc/enc.cpp:134]: For good practice of the API, please call `oe_verifier_initialize()` and/or `oe_attester_initialize()` and call only once.

File: samples/host_verify/host.c
Comment (wintersteiger) [reply samples/host_verify/host.c]: Done, thanks!

File: tests/tools/oecert/enc/enc.cpp
Comment (wintersteiger) [reply tests/tools/oecert/enc/enc.cpp:134]: Thanks for clarifying @yentsanglee. I moved the verifier initialization in host_verify so that it's only done once.

**Comment by wintersteiger** [comment]
bors try

**Comment by BRMcLaren** [comment]
bors try cancel

File: host/crypto/openssl/rsa.c
Comment (CodeMonkeyLeet) [inline host/crypto/openssl/rsa.c:461]: Nit: newline for EOF

**Review by CodeMonkeyLeet — APPROVED** [review]
The changes LGTM, thanks for the refactoring! A couple of minor edit comments inline, but I approving with the expectation that they're easy fixups. 

@yentsanglee The host_verify and oecert changes look okay to me, but can you review and approve since you have better context?

File: include/openenclave/internal/rsa.h
Comment (dthaler) [inline include/openenclave/internal/rsa.h:274]:
```suggestion
 * @param public_key[out] The key to be constructed.
```

File: tests/tools/oecert/host/host.cpp
Comment (dthaler) [inline tests/tools/oecert/host/host.cpp:305]: This file already includes <openenclave/host.h> which has (in lines 35-48) mappings to allow code to just use the C11 _s APIs and not have to use ifdefs in all these files.  Consider adding fopen_s to that list there and then just using line 302 here without ifdefs, and same for 346 etc.   fopen_s has been part of C11 (i.e., standard C since 2011) per https://en.cppreference.com/w/c/io/fopen

E.g.,
#define fopen_s(pfp, name, mode) *(pfp) = fopen((name), (mode))

**Review by dthaler — APPROVED** [review]

**Comment by yentsanglee** [comment]
> The changes LGTM, thanks for the refactoring! A couple of minor edit comments inline, but I approving with the expectation that they're easy fixups.
> 
> @yentsanglee The host_verify and oecert changes look okay to me, but can you review and approve since you have better context?

The host_verify and oecert changes look good to me.

**Review by yentsanglee — APPROVED** [review]
The changes in host_verify and oecert look good to me.

File: enclave/crypto/rsa.c
Comment (wintersteiger) [reply enclave/crypto/rsa.c:361]: I blindly assumed clang-format would do that for me, but apparently it doesn't support that. However, VSCode supports it and it's fine doing that on top of clang-format's changes.

File: host/crypto/bcrypt/rsa.c
Comment (wintersteiger) [reply host/crypto/bcrypt/rsa.c]: Sure, fixed.

Comment (wintersteiger) [reply host/crypto/bcrypt/rsa.c]: I really shouldn't use `OE_UNEXPECTED` for this type of error!

File: tests/tools/oecert/host/host.cpp
Comment (wintersteiger) [reply tests/tools/oecert/host/host.cpp]: Thanks for the pointer, I wasn't aware of those macros yet. All fixed now.

**Comment by wintersteiger** [comment]
bors try

**Comment by wintersteiger** [comment]
bors try

**Comment by wintersteiger** [comment]
bors try

**Comment by wintersteiger** [comment]
bors try

**Comment by wintersteiger** [comment]
bors try

**Comment by wintersteiger** [comment]
bors try

**Comment by wintersteiger** [comment]
bors try

**Comment by wintersteiger** [comment]
bors try

**Comment by wintersteiger** [comment]
I think this one's ready to merge, could I ask one of the approvers to r+ it? Thanks!

**Pull request merged**
